src/librustc_infer/traits/wf.rs

```diff
@@ -604,11 +605,14 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
         substs: SubstsRef<'tcx>,
     ) -> Vec<traits::PredicateObligation<'tcx>> {
         let predicates = self.infcx.tcx.predicates_of(def_id).instantiate(self.infcx.tcx, substs);
-        let cause = self.cause(traits::ItemObligation(def_id));
         predicates
             .predicates
             .into_iter()
-            .map(|pred| traits::Obligation::new(cause.clone(), self.param_env, pred))
+            .zip(predicates.spans.into_iter())
+            .map(|(pred, span)| {
+                let cause = self.cause(traits::ItemObligation(def_id, Some(span)));
+                traits::Obligation::new(cause, self.param_env, pred)
+            })
```

Comment:
And here it is, this (mostly) self-contained change is what I suspect is responsible for most of the test changes, I would love to have this landed ASAP (assuming sticking the `Span` into `ItemObligation` is the best way to go about this, I'd defer to @nikomatsakis on that).

Reply:
Just to be clear, this change is the one that precipitated the shift to keeping `PredicateObligation`s instead of `Predicate`s.